tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc

```diff
@@ -729,6 +732,192 @@ class ConvertEinsum : public OpConverterBase<ConvertEinsum> {
+    LabelCounts output_label_counts;
+    absl::InlinedVector<bool, 2> input_has_ellipsis;
+    bool output_has_ellipsis;
+    VLOG(2) << "Parsing equation " << eq;
+    TF_RETURN_IF_ERROR(ParseEinsumEquation(
+        eq, &input_labels, &output_labels, &label_types, &input_label_counts,
+        &output_label_counts, &input_has_ellipsis, &output_has_ellipsis));
+
+    Status unimplemented =
+        errors::Unimplemented("No conversion for einsum equation.");
+    if (input_has_ellipsis[0] || (inputs.size() > 1 && input_has_ellipsis[1]) ||
+        output_has_ellipsis) {
+      VLOG(2) << "Ellipsis not yet supported";
+      return unimplemented;
+    }
+    for (int i = 0; i < input_label_counts.size(); i++) {
+      for (int k = 0; k < input_label_counts[i].size(); k++) {
+        if (input_label_counts[i][k] > 1) {
+          VLOG(2) << "Diagonal operation or reduction not yet supported";
+          return unimplemented;
+        }
+      }
+    }
+    bool has_dynamic_shape = false;
+    for (const auto& input : inputs) {
+      VLOG(2) << "input dims " << DebugString(input.GetTrtDims());
+      has_dynamic_shape |= !HasStaticShape(input.GetTrtDims());
+    }
+    bool has_out_idx =
+        std::reduce(output_label_counts.begin(), output_label_counts.end(),
+                    false, std::logical_or<int>());
+    for (int i = 0; i < output_label_counts.size(); i++)
+      VLOG(2) << "O[" << i << "]: " << output_label_counts[i];
+    VLOG(2) << "has dynamic shape " << has_dynamic_shape << ", has_out_idx "
+            << has_out_idx;
+    if (has_dynamic_shape && !has_out_idx) {
+      VLOG(2) << "Scalar output not allowed in dynamic shape mode";
+      return unimplemented;
+    }
+    // Check if it is a dot product
+    if (input_label_counts.size() == 2 && input_label_counts[0].size() == 1 &&
+        input_label_counts[0][0] == 1 && input_label_counts[1][0] == 1 &&
+        output_label_counts[0] == 0) {
+      VLOG(2) << "Dot product not supported";
+      return unimplemented;
+    }
+    // Check for outer product
+    if (input_label_counts.size() == 2 && output_label_counts.size() == 2 &&
+        output_label_counts[0] == 1 && output_label_counts[1] == 1) {
+      VLOG(2) << "Outer product not supported";
+      return unimplemented;
+    }
+    return Status::OK();
+  }
+
+  Status Validate() {
+    VLOG(2) << "Running validation using the new einsum "
+               "converter";
+    if (params_->use_implicit_batch) {
+      return errors::Unimplemented(
+          "Einsum converter requires dynamic shape mode");
+    }
+
+    for (const auto& input : params_->inputs) {
+      VLOG(2) << "input " << (input.is_tensor() ? "tensor" : "weight")
+              << ", dims " << DebugString(input.GetTrtDims());
+    }
+    StatusOr<std::string> eq = GetAttrValue<std::string>("equation");
+    TRT_ENSURE_OK(eq);
+
+    TF_RETURN_IF_ERROR(ValidateEinsumEquation(*eq));
+
+    equation = MakeLowerCase(*eq);
```

Comment:
Can you put a comment to explain why we need to rewrite the equation with lower cases?